src/ORM/DataQuery.php

```diff
@@ -723,6 +721,56 @@ public function conjunctiveGroup()
         return new DataQuery_SubGroup($this, 'AND', $clause);
     }
 
+    /**
+     * Adds a Common Table Expression (CTE), aka WITH clause.
+     *
+     * Use of this method should usually be within a conditional check against DB::get_conn()->supportsCteQueries().
+     *
+     * @param string $name The name of the WITH clause, which can be referenced in any queries UNIONed to the $query
+     * and in this query directly, as though it were a table name.
+     * @param string[] $cteFields Aliases for any columns selected in $query which can be referenced in any queries
+     * UNIONed to the $query and in this query directly, as though they were columns in a real table.
+     * NOTE: If $query is a DataQuery, then cteFields must be the names of real columns on that DataQuery's data class.
+     * @param string|string[] $onClause The "ON" clause (escaped SQL statement) for an INNER JOIN on the query.
+     * It can either be a full clause (like you would pass to {@link leftJoin()} or {@link innerJoin()}),
+     * or it can be an array mapping of the field(s) on the dataclass table that map with the field(s) on the CTE table
+     * e.g. ['ID' => 'cte_id']
+     * If you want to use another join type, leave this blank and call the appropriate join method.
+     */
+    public function with(string $name, DataQuery|SQLSelect $query, array $cteFields = [], string|array $onClause = '', bool $recursive = false): static
```

Comment:
I feel like we should require $query to only be a `DataQuery` here and not allow a `SQLSelect` for the sake of simplicity. I found the docs PRs which mention both `DataQuery` and `SQLSelect` a bit overwhelming and for the average person who may not even know about anything beyond `DataList` it's almost certainly too much.  For the sake of simplicity and comprehension if it's possible I think we should only expose/encourage people to use `DataQuery`

Reply:
> for the average person who may not even know about anything beyond `DataList` it's almost certainly too much

If we wanted those people to interact with this API, we would be adding an abstraction layer to `DataList`. I'm intentionally not putting an abstraction layer there because CTEs are a complex topic that can't be easily abstracted to a level that I'd be comfortable putting there. This is an intentionally fairly low level API.

> I feel like we should require $query to only be a DataQuery here and not allow a SQLSelect for the sake of simplicity. 

I strongly disagree with doing this. Writing `SQLSelect` queries that work exactly how you need them to work for complex scenarios is a lot easier than writing `DataQuery` queries. It is very likely that people (including us) will find they want to use an `SQLSelect` for the CTE portion of the query, and pass that in here to modify the output of a `DataQuery` result set.

If the documentation is confusing, we can resolve that by making the documentation as clear as possible. We shouldn't make the implementation objectively more difficult to use to resolve complex documentation.

Reply:
In that case would it make sense to go in the other direction and only allow SQLSelect here? I'm really not a fan of mixed params

Reply:
Just as there are scenarios where it would be easier to directly use `SQLSelect` for more complex CTEs, there are times when it would be easier to let a `DataQuery` handle most of the work for you for simpler CTEs, especially non-recursive ones.

You accepted `DataQuery|SQLSelect` for the union method signature which you reviewed and merged recently, I don't see how this would be any different.

Reply:
Ah OK, if you think it'll make stuff simpler overall then it makes sense